content/document.js

```diff
@@ -88,33 +88,41 @@ function extractLocale(folder) {
   return locale;
 }
 
-function saveFile(
-  filePath,
-  rawBody,
-  { slug, title, translation_of, tags, translation_of_original, original_slug }
-) {
-  if (slug.includes("#")) {
-    throw new Error("newSlug can not contain the '#' character");
-  }
-  const metadata = {
-    title,
-    slug,
-  };
-  if (tags) {
-    metadata.tags = tags;
+function saveFile(filePath, rawBody, metadata, frontMatterKeys = null) {
+  const requiredFrontMatterKeys = ["title", "slug"];
+  const optionalFrontMatterKeys = [
+    "tags",
+    "translation_of",
+    "translation_of_original",
+    "original_slug",
+  ];
```

Comment:
Let's nerd-out about this because I understand your point but I'm not convinced. 
But being inside the function, they're "naturally organized" by the scope of the function. 
An advantage of that is you can name them "sloppily". E.g. 

```js
function foo() {
  const keys = ['...']
  ...
}

function bar() {
  const keys = ['...']
  ...
}
```
makes it OK to call it just `keys`. If you were to put them outside the function scopes you'd have to come up with specific names. 


```js
const FOO_KEYS = ['...']

function foo() {
  ...
}

const BAR_KEYS = ['...']

function bar() {
  ...
}
```

Now, they're almost unnecessarily verbose. And if you decide to refactor `function foo` you'll have to remember to look outside the function because it's no longer self-contained. 

But having them inside the function means that that array needs to be re-created every single time the function is called. There might be some overheads with that but the function would have to be called extremely frequently and time-sensitively for it to matter, right?

I do prefer to use `const UPPER_CASE_THING = ...` when the constants are very specific and has a chance of telling and interesting story. E.g. 

```js

// The current rate is dependent on this website for these reasons
const TAX_RATE = 0.123

...


function calculateTaxCredit(amount) {
  ...
}
```

I think the context here it's nice that they're inside the function. It's often a bit weird when you create a variable but only refer to it once! But I think...

```js
const keys = [
  'one',
  'two',
  'three',
  'four',
];
for (const key of keys) {
  // ..
}
```
...reads better than...
```js
for (const key of [
  'one',
  'two',
  'three',
  'four',
]) {
  // ..
}
```
Partly it makes the code a bit more dumb and readable but also the variable *name* can actually "tell a useful story" which you'd have to use a code comment to explain.

Reply:
I do see that grouping benefit and I also think the performance cost is probably negligble of keeping it inside the closure (I can imagine v8 having some optimizations for those that are not mutated inside the fn).

What I like about moving them out-of-scope is that I then also don't expect them to be mutated inside of it, which is nice, because my brain has fewer things to keep in mind while reading the beast. Now to be fair, there are cases where we still mutate module-globals, but it's the minority of cases, so that's nice.
And yeah I guess the same could be achieved with capitalized letters.

I'm quite ambivalent about it, though my personal default is definitely pulling it out!